druid-shell/src/platform/windows/window.rs

```diff
@@ -454,11 +596,97 @@ impl WndProc for MyWndProc {
                 }
                 Some(0)
             },
+            WM_DPICHANGED => unsafe {
+                let x = HIWORD(wparam as u32) as f64 / SCALE_TARGET_DPI;
+                let y = LOWORD(wparam as u32) as f64 / SCALE_TARGET_DPI;
+                let scale = Scale::new(x, y);
+                self.set_scale(scale);
+                let rect: *mut RECT = lparam as *mut RECT;
+                SetWindowPos(hwnd, HWND_TOPMOST, (*rect).left, (*rect).top, (*rect).right - (*rect).left, (*rect).bottom - (*rect).top, SWP_NOZORDER | SWP_FRAMECHANGED | SWP_DRAWFRAME);
+                if let Ok(mut s) = self.state.try_borrow_mut() {
+                    let s = s.as_mut().unwrap();
+                    if s.dcomp_state.is_some() {
+                        let scale = self.scale();
+                        let rt = paint::create_render_target(&self.d2d_factory, hwnd, scale);
+                        s.render_target = rt.ok();
+                        {
+                            let rect_dp = self.area().size_dp().to_rect();
+                            s.handler.rebuild_resources();
+                            s.render(
+                                &self.d2d_factory,
+                                &self.dwrite_factory,
+                                &self.handle,
+                                rect_dp,
+                            );
+                        }
+
+                        if let Some(ref mut ds) = s.dcomp_state {
+                            let _ = ds.dcomp_target.clear_root();
+                            let _ = ds.dcomp_device.commit();
+                            ds.sizing = true;
+                        }
+                    }
+                } else {
+                    self.log_dropped_msg(hwnd, msg, wparam, lparam);
+                }
+                Some(0)
+            },
+            WM_NCCALCSIZE => unsafe {
+                // Hack to get rid of caption but keeping the borders created by it.
+                let style = GetWindowLongPtrW(hwnd, GWL_STYLE) as u32;
+                if style == 0 {
+                    warn!(
+                        "failed to get window style: {}",
+                        Error::Hr(HRESULT_FROM_WIN32(GetLastError()))
+                    );
+                    return Some(0);
+                }
+                if !self.has_titlebar() && (style & WS_CAPTION) != 0 {
+                    let s: *mut NCCALCSIZE_PARAMS = lparam as *mut NCCALCSIZE_PARAMS;
+                    if let Some(mut s) = s.as_mut() {
+                        s.rgrc[0].top -= (31.0 * self.scale().x()) as i32;
```

Comment:
These hard-coded numbers seem scary to me. I would be more comfortable if they were at least moved to `const` with docstring explaining where they came from.